### ISSUE TITLE
tracking.max_tracking memory bloating

### ISSUE THREAD

**Issue by researcherofreality** [talmolab/sleap#2211]
Edit: opened a discussion page for questions, seems to be a more appropriate place. The only issue that I would like to point out is the memory leak occurring for tracking models other than `simple` for my system/sleap config, and also occurring for the `tracking.max_tracking` option.

Hi,

First off, great software! It's been fun getting into.

I am trying to get sleap running on my experiment, where we have a single animal in a large arena. From https://github.com/talmolab/sleap/discussions/1732 it seems to make sense to use a top-down model, which I went forward with, with quite good results. There's a few things I would like to ask help on, and point out. 

Toy problem CLI I use is the following 
```
sleap-track "/path/to/video.mp4" 
-m "/path/to/centroid_dir" -m "/path/to/centered_instance_dir" 
--frames 18500-34000 -o "./sleap_output/output.slp" 
--open-in-gui  
--max_instances 1 
--tracking.tracker simple 
--tracking.max_tracking 1 
--tracking.similarity instance 
```

Explanations & observations:
1. I added the max_instances 1 and max_tracking 1 option to avoid having multiple instances & tracks, hopefully helping the model to infer and avoid fixing the tracks afterwards
2. I added the simple tracker with similarity instance option to hopefully get more consistency in the inferred keypoints. Sometimes the animal moves very very slightly but a keypoint or two gets lost on the neural recording tether. I picked simple because I was observing memory leakage/bloating when using others, similar to  https://github.com/talmolab/sleap/issues/2081 
3. It seems that the memory leak observed in the above issue also occurs for the tracking.max_tracking option, even when using the simple tracker. When I run the same command without it, the memory stays flat. 

My question is, even when using the tracker some keypoints seem to jump on the neural recording tether sometimes. My assumption based on https://sleap.ai/develop/guides/proofreading.html#tracking-method-details is that using the tracker with instance similarity option should avoid this.: _" “instance” measures similarity by looking at the distances between corresponding nodes in the instances, normalized by the number of valid nodes in the candidate instance. "_
Should this work? Or should I try something else? 

Also, whats the easiest way to have only 1 track and 1 instance? Any considerations? Would like minimal post-processing for it.

Ill add some system and software configs as comment.

Thank you!

**Comment by researcherofreality** [comment]
```
memory     256GiB System memory
processor  AMD Ryzen Threadripper PRO 5975WX 32-Cores

 nvidia-smi
Fri Jun 27 16:11:39 2025
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.35.02              Driver Version: 560.94         CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4070 ...    On  |   00000000:41:00.0  On |                  N/A |
| 38%   68C    P2            190W /  220W |   11338MiB /  12282MiB |     86%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                 
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A       390      G   /Xwayland                                   N/A      |
|    0   N/A  N/A    641281    C+G   /sleap-label                                N/A      |
+-----------------------------------------------------------------------------------------+
```

Installed sleap using 
`conda create -y -n sleap -c conda-forge -c nvidia -c sleap/label/dev -c sleap -c anaconda sleap=1.4.1
`

switched to nvidia-cudnn-cu11==8.6.0.163 because of ncompatibility 

SLEAP: 1.4.1a2
TensorFlow: 2.7.0
Numpy: 1.21.6
Python: 3.7.12
OS: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-debian-trixie-sid

**Comment by researcherofreality** [comment]
I found that setting max instance to 1 obviously constrains the training and inference to 1 instance. Also I think the command in https://sleap.ai/develop/guides/proofreading.html 
`python -m sleap.info.trackcleaner path/to/predictions.h5 -c 1` should help with the tracks post-hoc.

However, I am still wondering about whether there is a way to help the model during inference to find the keypoints, i.e. some keypoint velocity or inter-keypoint velocity kind of metric which could constrain the jumping around of certain keypoints.

**Comment by researcherofreality** [comment]
I'll open a discussion page for my questions, seems to be a more appropriate place. The only issue that I would like to point out is the memory leak occurring for tracking models other than `simple` for my system/sleap config, and also occurring for the `tracking.max_tracking` option.

**Comment by talmo** [comment]
Thanks for flagging this, and sorry for the slow follow-up.

The memory growth with the non-`simple` trackers was a known issue in the legacy TensorFlow tracking stack (same as #2081). That entire code path (`sleap/nn/tracking.py`) was removed in the 1.5 PyTorch migration — the new tracker uses bounded queues, so upgrading to [1.6.3](https://docs.sleap.ai/latest/installation/) should resolve it.

Your single-animal setup and keypoint-jumping questions are covered over in #2214. Closing this as completed — if you still see memory bloat on the current version, please open a fresh issue with a `top`/`mprof` profile and we'll dig in.
